### PR TITLE
Results table observer for muons

### PR DESCRIPTION
**Pull request by AnthonyLim23**
**Description of work.**

**Merge after https://github.com/mantidproject/mantid/pull/33743**

Users wanted to be able to change the values in a results table and for the values to automatically be updated in the GUI. 

**To test:**

Before opening Mantid turn on the model analysis and plotting (option 2) feature flag as described in https://github.com/mantidproject/mantid/pull/33712

Open mantid
Open muon analysis
Load MUSR 62260
Go to the fitting tab and add a function (doesnt matter which)
Go to sequential fitting tab and press fit all
Go to the results table tab
Select run number in the top table, then click the "output results" button 
Change the table name to "run" and press "output results"
Change the table name to "new" and press "output results"
Go to the model fitting tab
Notice that the plot is just bunch of markers in a vertical line
Open the "new" table from the ADS
Change one of the run numbers to 62249
Notice that the plot updates
Deleting "run" from the ADS will update the results table combo box so it will only contain "new" and "ResultsTable"

Fixes #33763. 

There is no release note for this as it is part of the new "model analysis" 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/contexts/muon_context.py [hunk 1]
```diff
@@ -536,14 +537,16 @@ def remove_workspace(self, workspace):
         # required as the renameHandler returns a name instead of a workspace.
         if isinstance(workspace, str):
             workspace_name = workspace
+        if isinstance(workspace, TableWorkspace):
```
Comment (AnthonyLim23) [inline qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/contexts/muon_context.py:540]:
```suggestion
        elif isinstance(workspace, TableWorkspace):
```

**Comment by AnthonyLim23** [comment]
@Harrietbrown could you please keep an eye on this while I am away?

**Review by SilkeSchomann — CHANGES_REQUESTED** [review]
It does work as expected but it looks as if a comment in the code needs to be cleaned up.

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/contexts/results_context.py [hunk 2]
```diff
@@ -5,13 +5,31 @@
...
 
 
 class ResultsContext:
 
     def __init__(self):
         self._result_table_names: list = []
 
+        self.ADS_observer = MuonADSObserver(self.remove,
+                                            self.clear,
+                                             self.replaced) # need to add unsubscribe to a close method in the widget
```
[the remark this answers is not in the document]
Comment (Harrietbrown) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/contexts/results_context.py]: Anthony did mention a issue with subscribers not being removed, that was a seperate problem so this comment was likly intended as a TODO. I've removed the comment.

**Review by SilkeSchomann — APPROVED** [review]
Happy to approve this now after the changes.

**Pull request merged**